src/test/groovy/com/pretz/parkingmanager/mapper/ParkingSessionMapperSpec.groovy

```diff
@@ -0,0 +1,147 @@
+        testModelMapper.addConverter(new ParkingSessionToParkingMeterResponseDTOConverter())
+        parkingSessionMapper = new ParkingSessionMapper(testModelMapper)
+
+        parkingStartDTOBuilder = ParkingStartDTO.builder()
+                .vehicleId("ABC1234")
+                .parkingRateId(1)
+
+        parkingSessionBuilder = ParkingSession.builder()
+                .vehicleId("EWE4567")
+                .id(1)
+    }
+
+    @Unroll
+    def "mapper should map ParkingStartDTO to ParkingSession object with vehicleId and parkingRate fields initalized and other fields null"() {
+
+        given:
+        ParkingStartDTO testParkingStartDTO = parkingStartDTOBuilder
+                .vehicleId(generatedValues.first)
+                .parkingRateId(generatedValues.second)
+                .build()
+
+        when:
+        ParkingSession testParkingSession = parkingSessionMapper.fromParkingStartDTO(testParkingStartDTO)
+
+        then:
+        testParkingSession.getVehicleId() == testParkingStartDTO.getVehicleId()
+        testParkingSession.getParkingRate() != null
+        testParkingSession.getId() == 0
+        testParkingSession.getStartTime() == null
+        testParkingSession.getStopTime() == null
+
+        where:
+        generatedValues << somePairs(strings(), longs(0, 100))
+    }
+
+    def "mapper should map ParkingStartDTO to ParkingSession object with REGULAR parking rate"() {
+
+        given:
+        ParkingStartDTO testParkingStartDTO = parkingStartDTOBuilder
+                .parkingRateId(1)
+                .build()
+
+        when:
+        ParkingSession testParkingSession = parkingSessionMapper.fromParkingStartDTO(testParkingStartDTO)
+
+        then:
+        testParkingSession.getParkingRate() == ParkingRate.REGULAR
+    }
+
+    def "mapper should map ParkingStartDTO to ParkingSession object with DISABLED parking rate"() {
+
+        given:
+        ParkingStartDTO testParkingStartDTO = parkingStartDTOBuilder
+                .parkingRateId(2)
+                .build()
+
+        when:
+        ParkingSession testParkingSession = parkingSessionMapper.fromParkingStartDTO(testParkingStartDTO)
+
+        then:
+        testParkingSession.getParkingRate() == ParkingRate.DISABLED
+    }
+
+    def "mapper should map ParkingStartDTO to ParkingSession object with NONE parking rate"() {
+
+        given:
+        ParkingStartDTO testParkingStartDTO = parkingStartDTOBuilder
+                .parkingRateId(0)
+                .build()
+
+        when:
+        ParkingSession testParkingSession = parkingSessionMapper.fromParkingStartDTO(testParkingStartDTO)
+
+        then:
+        testParkingSession.getParkingRate() == ParkingRate.NONE
+    }
+
+    def "mapper should map ParkingSession to ParkingMeterResponseDTO with timestamp calculated from start time if stop time is null"() {
+
+        given:
+        Timestamp startTimestamp = Timestamp.from(Instant.now())
```

Comment:
czy korzystanie z czasu w ten sposób nie jest trudne? Czy znasz lepszy sposób testowania z użyciem czasu, który da się lepiej kontrolować?

Reply:
Fakt, korzystanie z czegoś takiego powoduje że testy są nie do końca deterministyczne i o ile tutaj raczej to nie powinno przeszkadzać, to nie jest to dobra praktyka. Lepszą byłoby inicjalizowanie timestampów stałymi wartościami.